dasharo-stability/capsule-update.robot

```diff
@@ -116,7 +115,7 @@ CUP160.001 Verifying BIOS Settings Persistence After Update - PART 2
 CUP170.001 Verifying UUID (Ubuntu)
     [Documentation]    Check if UUID didn't change after Capsule Update.
     ${tmp}=    Get Variable Value    $UPDATED_UUID
-    IF    $tmp is None
+    IF    '${tmp}' == 'None'
```

Comment:
`$tmp is None` doesnt work

Reply:
Indeed, `Evaluating IF condition failed: Evaluating expression '$tmp is None' failed: SyntaxError: invalid syntax (<string>, line 1)`. Interestingly, this exact syntax is described in https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#special-variable-syntax:

> For example, if the variable value can be either a string or Python `None`, quoting like `'${var}'` is needed because otherwise strings do not work, but then `None` is interpreted to be a string as well. Luckily there is an easy solution to these problems discussed in this section.

`${tmp} is None` works fine for UUID and serial, but not for logo:
`Evaluating IF condition failed: Evaluating expression 'bf6cd62f794e18db7c18816ba38401204a9a2bf2532f201f3bfaee67608b8511  /sys/firmware/acpi/bgrt/image is None' failed: NameError: name 'bf6cd62f794e18db7c18816ba38401204a9a2bf2532f201f3bfaee67608b8511' is not defined nor importable as module`

Using `'${tmp}' is None` also won't work, because `'None'` isn't `None`, same for `'$tmp' is None`.

Doing comparison in inline Python (`${{$tmp is None}}` etc.) doesn't change anything, neither does passing it through `Evaluate`.

I guess this really is the best we can get from RF, our expectations exceeded the tool capabilities. It doesn't help that the documentation is wrong on the original version being a solution to this exact problem.

Reply:
@SergiiDmytruk This thread is blocking the merge (specifically, rebase -> testing on MSI -> and then merging). Is this explanation enough for you?

Reply:
Sure, it was just a question because the docs claim it should work.

Reply:
@krystian-hebel, do you have a concrete example where `$tmp is None` doesn't work?

Reply:
@pekkaklarck thanks for showing interest, it was unexpected. Out of curiosity, how did you find the mention of this problem here? :slightly_smiling_face: 

Simplest example:

```
*** Test Cases ***
TEST001 Tests suite variable
    Set Suite Variable    ${MY_VAR}    Some value
    IF    $MY_VAR is None
        Log To Console    MY_VAR is None
    ELSE
        Log To Console    MY_VAR is $MY_VAR
    END
```

Result:

```
$ robot --version
Robot Framework 5.0 (Python 3.12.7 on linux)
$ robot test.robot 
==============================================================================
Test                                                                          
==============================================================================
TEST001 Tests suite variable                                          | FAIL |
Evaluating IF condition failed: Evaluating expression '$MY_VAR is None' failed: SyntaxError: invalid syntax (<string>, line 1)
------------------------------------------------------------------------------
Test                                                                  | FAIL |
1 test, 0 passed, 1 failed
==============================================================================
```

Reply:
I am afraid it might be fixed in newer versions:

```
python3 -m robot -h
Robot Framework -- A generic automation framework

Version:  7.1.1 (Python 3.12.7 on linux)
```

```
(venv) macpijan in /tmp/pp λ python3 -m robot ./test.robot 
==============================================================================
Test                                                                          
==============================================================================
TEST001 Tests suite variable                                          .MY_VAR is $MY_VAR
TEST001 Tests suite variable                                          | PASS |
------------------------------------------------------------------------------
Test                                                                  | PASS |
1 test, 1 passed, 0 failed
==============================================================================
Output:  /tmp/pp/output.xml
Log:     /tmp/pp/log.html
Report:  /tmp/pp/report.html
```

We're still on the v5.0, could potentially upgrade (with the necessity to rebase some patches). The final goal is to use the upstream version, once we resolve the migration of  telnet lib https://github.com/MarketSquare/robotframework-telnet-library

Reply:
The problem with `$MY_VAR is None` with RF 5.0 is that this release isn't fully compatible with Python 3.12 that you used. Subtle changes to Python's tokenizer broke the `$var` syntax as you can read from this issue:
https://github.com/robotframework/robotframework/issues/4771

Your options include upgrading to RF 6.1 or newer or downgrading to Python 3.11.